src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs

```diff
@@ -1,93 +1,597 @@
+    {
+        QuicStream stream = new QuicStream(new Implementations.MsQuic.MsQuicStream(_state, _handle, data.Stream, data.Flags));
+        if (!_acceptQueue.Writer.TryWrite(stream))
+        {
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Error(this, $"{this} Unable to enqueue incoming stream {stream}");
+            }
+            stream.Dispose();
+            return QUIC_STATUS_SUCCESS;
+        }
+        return QUIC_STATUS_SUCCESS;
+    }
+    private unsafe int HandleEventPeerCertificateReceived(ref PEER_CERTIFICATE_RECEIVED_DATA data)
+    {
+        try
+        {
+            return _sslConnectionOptions.ValidateCertificate((QUIC_BUFFER*)data.Certificate, (QUIC_BUFFER*)data.Chain, out _remoteCertificate);
+        }
+        catch (Exception ex)
+        {
+            _connectedTcs.TrySetException(ex);
+            return QUIC_STATUS_HANDSHAKE_FAILURE;
+        }
+    }
+
+    private unsafe int HandleConnectionEvent(ref QUIC_CONNECTION_EVENT connectionEvent)
+        => connectionEvent.Type switch
+        {
+            QUIC_CONNECTION_EVENT_TYPE.CONNECTED => HandleEventConnected(ref connectionEvent.CONNECTED),
+            QUIC_CONNECTION_EVENT_TYPE.SHUTDOWN_INITIATED_BY_TRANSPORT => HandleEventShutdownInitiatedByTransport(ref connectionEvent.SHUTDOWN_INITIATED_BY_TRANSPORT),
+            QUIC_CONNECTION_EVENT_TYPE.SHUTDOWN_INITIATED_BY_PEER => HandleEventShutdownInitiatedByPeer(ref connectionEvent.SHUTDOWN_INITIATED_BY_PEER),
+            QUIC_CONNECTION_EVENT_TYPE.SHUTDOWN_COMPLETE => HandleEventShutdownComplete(ref connectionEvent.SHUTDOWN_COMPLETE),
+            QUIC_CONNECTION_EVENT_TYPE.LOCAL_ADDRESS_CHANGED => HandleEventLocalAddressChanged(ref connectionEvent.LOCAL_ADDRESS_CHANGED),
+            QUIC_CONNECTION_EVENT_TYPE.PEER_ADDRESS_CHANGED => HandleEventPeerAddressChanged(ref connectionEvent.PEER_ADDRESS_CHANGED),
+            QUIC_CONNECTION_EVENT_TYPE.PEER_STREAM_STARTED => HandleEventPeerStreamStarted(ref connectionEvent.PEER_STREAM_STARTED),
+            QUIC_CONNECTION_EVENT_TYPE.PEER_CERTIFICATE_RECEIVED => HandleEventPeerCertificateReceived(ref connectionEvent.PEER_CERTIFICATE_RECEIVED),
+            _ => QUIC_STATUS_SUCCESS
+        };
+
+#pragma warning disable CS3016
+    [UnmanagedCallersOnly(CallConvs = new Type[] { typeof(CallConvCdecl) })]
+#pragma warning restore CS3016
+    private static unsafe int NativeCallback(QUIC_HANDLE* connection, void* context, QUIC_CONNECTION_EVENT* connectionEvent)
+    {
+        GCHandle stateHandle = GCHandle.FromIntPtr((IntPtr)context);
+
+        // Check if the instance hasn't been collected.
+        if (!stateHandle.IsAllocated || stateHandle.Target is not QuicConnection instance)
+        {
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Error(null, $"Received event {connectionEvent->Type}");
```

Comment:
Should we elaborate in the message? and explain that it's not allocated anymore, or it's some strange instance (how could that be??)

Reply:
Fixed all occurrences, also fixed in stream.